pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -527,21 +527,33 @@ func (nim *nodeInfoManager) installDriverToCSINode(
 		topologyKeys.Insert(k)
 	}
 
+	var maxAllocatableCount int32
+	if maxAttachLimit > 0 {
+		if maxAttachLimit > math.MaxInt32 {
+			klog.Warningf("Exceeded max supported attach limit value, truncating it to %d", math.MaxInt32)
+			maxAttachLimit = math.MaxInt32
+		}
+		maxAllocatableCount = int32(maxAttachLimit)
+	} else {
+		klog.Errorf("Invalid attach limit value %d cannot be added to CSINode object for %q", maxAttachLimit, driverName)
+	}
+
 	specModified := true
 	// Clone driver list, omitting the driver that matches the given driverName
 	newDriverSpecs := []storagev1.CSINodeDriver{}
 	for _, driverInfoSpec := range nodeInfo.Spec.Drivers {
 		if driverInfoSpec.Name == driverName {
 			if driverInfoSpec.NodeID == driverNodeID &&
-				sets.NewString(driverInfoSpec.TopologyKeys...).Equal(topologyKeys) {
+				sets.NewString(driverInfoSpec.TopologyKeys...).Equal(topologyKeys) &&
+				((maxAllocatableCount == 0 && (driverInfoSpec.Allocatable == nil || driverInfoSpec.Allocatable.Count == nil)) ||
+					(driverInfoSpec.Allocatable != nil && driverInfoSpec.Allocatable.Count != nil && *driverInfoSpec.Allocatable.Count == maxAllocatableCount)) {
```

Comment:
nit: This is a fairly complex if statement. Can we explore ways to make this easier to read? One way I was thinking was booleans for each condition:
```go
topologyKeysMatch := sets.NewString(driverInfoSpec.TopologyKeys...).Equal(topologyKeys)
zeroMaxAllocMatch := maxAllocatableCount == 0 && (driverInfoSpec.Allocatable == nil || driverInfoSpec.Allocatable.Count == nil)
nonZeroMaxAllocMatch := driverInfoSpec.Allocatable != nil && driverInfoSpec.Allocatable.Count != nil && *driverInfoSpec.Allocatable.Count == maxAllocatableCount

if driverInfoSpec.NodeID == driverNodeID && topologyKeysMatch && (zeroMaxAllocMatch || nonZeroMaxAllocMatch) {
  ...
}
```

Reply:
Thanks for your kindly tips, could you please take another look?